onnxruntime/core/optimizer/concat_slice_elimination.cc

```diff
@@ -0,0 +1,235 @@
+      const NodeArg* input = get_input_if_exists(input_idx);
+      return input ? graph_utils::GetConstantInitializer(graph, input->Name()) : nullptr;
+    };
+
+    auto get_initializer_data =
+        [&graph](const ONNX_NAMESPACE::TensorProto* initializer) -> std::vector<int64_t> {
+      Initializer init(*initializer, graph.ModelPath());
+      if (initializer->data_type() == ONNX_NAMESPACE::TensorProto::INT32) {
+        int32_t* init_data = init.data<int32_t>();
+        return std::vector<int64_t>(init_data, init_data + init.size());
+      } else if (initializer->data_type() == ONNX_NAMESPACE::TensorProto::INT64) {
+        int64_t* init_data = init.data<int64_t>();
+        return std::vector<int64_t>(init_data, init_data + init.size());
+      }
+      return {};
+    };
+
+    // Starts and ends inputs have to exist, be constant, and be of the same size.
+    const ONNX_NAMESPACE::TensorProto* starts_init = get_initializer_if_constant(1);
+    const ONNX_NAMESPACE::TensorProto* ends_init = get_initializer_if_constant(2);
+    if (starts_init && ends_init) {
+      starts = get_initializer_data(starts_init);
+      ends = get_initializer_data(ends_init);
+
+      if (starts.size() == 0 || ends.size() == 0 || starts.size() != ends.size()) {
+        return false;
+      }
+
+      // If axes input exists, it should be constant and of the same size as starts/ends.
+      if (get_input_if_exists(3)) {
+        const ONNX_NAMESPACE::TensorProto* axes_init = get_initializer_if_constant(3);
+        if (!axes_init || axes_init->dims_size() != 1 ||
+            static_cast<size_t>(axes_init->dims().Get(0)) != starts.size()) {
+          return false;
+        }
+        axes = get_initializer_data(axes_init);
+
+        // If steps input exists, it should be constant and all values should be 1.
+        if (get_input_if_exists(4)) {
+          const ONNX_NAMESPACE::TensorProto* steps_init = get_initializer_if_constant(4);
+          if (!steps_init) {
+            return false;
+          }
+          steps = get_initializer_data(steps_init);
+          if (steps.size() != starts.size()) {
+            return false;
+          }
+          for (int64_t step : steps) {
+            if (step != 1) {
+              return false;
+            }
+          }
+        }
+      }
+    } else {
+      // Should be unreachable, but just to be safe in case a new op version is added.
+      return false;
+    }
+  } else {
+    return false;
+  }
+  return true;
+}
+
+/**
+Apply Concat Slice Elimination transform. This transform removes the redundant
+Concat + Slice pattern in the application of the attention bias.
+
+Before transform:
+q_bias              -- >Slice [0, q] ---> Add_q  
+      \            | 
+k_bias-- Concat --> -->Slice [q, q+k]---> Add_k
+      /            | 
+v_bias              -->Slice [q+k, :]---> Add_v
+
+After transform:
+q_bias ---> Add_q  
+      
+k_bias ---> Add_k
+
+v_bias ---> Add_v
+
+*/
```

Comment:
Overall, i think the pattern here is too specific for QKV... 
The same logic can be applied to Concat with >1 inputs...